File: unity/AIArenaSDK/Runtime/Inference/FallbackBehaviourAI.cs
using UnityEngine;
using AIArena.SDK.Agents;

namespace AIArena.SDK.Inference
{
    /// <summary>
    /// Heuristic fallback AI. Used when the inference API times out or is unavailable.
    /// Decision logic is driven by the agent's AgentBehaviourState trait weights.
    /// </summary>
    public class FallbackBehaviourAI : MonoBehaviour
    {
        private AgentBehaviourState _behaviourState;

        public void Configure(AgentBehaviourState state)
        {
            _behaviourState = state;
        }

        /// <summary>Select an action based on trait-weighted heuristics.</summary>
        public AgentAction GetAction(GameState gameState)
        {
            if (_behaviourState == null || gameState.AvailableActions == null || gameState.AvailableActions.Length == 0)
            {
                return new AgentAction { ActionType = "IDLE", Confidence = 0.1f, Source = "FALLBACK" };
            }

            float healthPercent = gameState.Health / gameState.MaxHealth;
            string selectedAction;

            // If critically low health, consider fleeing
            if (healthPercent < _behaviourState.CautiousHealthThreshold)
            {
                float fleeProbability = _behaviourState.FleeWeight * (1f - healthPercent);
                if (Random.value < fleeProbability)
                {
                    selectedAction = FindAction(gameState.AvailableActions, "FLEE", "DODGE", "RETREAT");
                    if (selectedAction != null)
                        return MakeAction(selectedAction, 0.7f);
                }
            }

            // Otherwise use weighted random selection
            float roll = Random.value;
            float cumulative = 0f;

            cumulative += _behaviourState.AttackWeight;
            if (roll < cumulative)
            {
                selectedAction = FindAction(gameState.AvailableActions, "ATTACK", "ABILITY", "STRIKE");
                if (selectedAction != null) return MakeAction(selectedAction, 0.6f);
            }

            cumulative += _behaviourState.DefendWeight;
            if (roll < cumulative)
            {
                selectedAction = FindAction(gameState.AvailableActions, "DEFEND", "BLOCK", "SHIELD");
                if (selectedAction != null) return MakeAction(selectedAction, 0.6f);
            }

            cumulative += _behaviourState.SupportWeight;
            if (roll < cumulative)
            {
                selectedAction = FindAction(gameState.AvailableActions, "HEAL", "BUFF", "SUPPORT");
                if (selectedAction != null) return MakeAction(selectedAction, 0.5f);
            }

            // Default: pick the first available action
            return MakeAction(gameState.AvailableActions[0], 0.3f);
        }

        private string FindAction(string[] available, params string[] keywords)
        {
            foreach (var action in available)
            {
                foreach (var kw in keywords)
                {
                    if (action.ToUpper().Contains(kw.ToUpper())) return action;
                }
            }
            return null;
        }

        private AgentAction MakeAction(string actionType, float confidence) =>
            new AgentAction { ActionType = actionType, Confidence = confidence, Source = "FALLBACK" };
    }
}

File: unreal/AIArenaSDK/Source/AIArenaSDK/Private/Inference/FallbackBehaviourAI.cpp
#include "Inference/FallbackBehaviourAI.h"

void UAIArenaFallbackBehaviourAI::Configure(const FAIArenaAgentTraits& Traits)
{
    CachedTraits = Traits;
}

FAIArenaAgentAction UAIArenaFallbackBehaviourAI::GetAction(
    const FAIArenaGameState& GameState) const
{
    FAIArenaAgentAction Action;
    Action.Source     = TEXT("FALLBACK");
    Action.Confidence = 0.2f;

    const float HealthPct = GameState.MaxHealth > 0.f
                              ? GameState.Health / GameState.MaxHealth
                              : 1.f;

    // ── Trait-weighted decision tree (mirrors server-side heuristic) ──────────
    if (HealthPct < 0.25f)
    {
        // Critical health — flee regardless of aggression
        Action.ActionType = TEXT("FLEE");
    }
    else if (CachedTraits.Aggression > 0.6f &&
             !GameState.NearbyEnemies.IsEmpty())
    {
        // High aggression + enemies in range → attack nearest
        Action.ActionType    = TEXT("ATTACK");
        Action.TargetId      = GameState.NearbyEnemies[0].Id;
        Action.TargetPosition = GameState.NearbyEnemies[0].Position;
        Action.Confidence    = CachedTraits.Aggression * 0.4f; // still low — it's a fallback
    }
    else if (CachedTraits.RiskTolerance < 0.4f)
    {
        Action.ActionType = TEXT("DEFEND");
    }
    else
    {
        Action.ActionType = TEXT("IDLE");
    }

    return Action;
}
